src/borrow_tracker/tree_borrows/tree.rs

```diff
@@ -729,38 +729,42 @@ impl Tree {
     /// whereas they were not when `parent` was still there.
     fn remove_useless_children(&mut self, root: UniIndex, live: &FxHashSet<BorTag>) {
         // To avoid stack overflows, we roll our own stack.
-        // Each element in the stack consists of the current tag, and the number of the next child.
+        // Each element in the stack consists of the current tag, and the number of the
+        // next child to be processed.
 
-        // The other functions are written using the `TreeVisitorStack`, which does not work
+        // The other functions are written using the `TreeVisitorStack`, but that does not work here
         // since we need to 1) do a post-traversal and 2) remove nodes from the tree.
         // Since we do a post-traversal (by deleting nodes only after handling all children),
         // we also need to be a bit smarter than "pop node, push all children."
         let mut stack = vec![(root, 0)];
+        let mut children_of_node = SmallVec::new();
         while let Some((tag, nth_child)) = stack.last_mut() {
             let node = self.nodes.get(*tag).unwrap();
-            if *nth_child == node.children.len() {
-                // We have visited all children. Let's remove those that are useless.
-                // Split the children into useful and useless.
-                // FIXME: this copies new_children, even if nothing gets removed.
-                // A 2-pass approach could be more efficient. It could consist of
-                // 1. traverse the Tree, collect all useless tags in a Vec
-                // 2. traverse the Vec, remove all tags previously selected
-                let (useless, remaining_children) =
-                    node.children.iter().copied().partition(|x| self.is_useless(*x, live));
-                // Assign the new children.
-                // This might make this node useless as well, but removal is the parent's job.
-                self.nodes.get_mut(*tag).unwrap().children = remaining_children;
-                // Next, ensure all useless children are properly unregistered everywhere.
-                for idx in useless {
-                    // Note: In the rest of this comment, "this node" means that at idx.
+            if *nth_child < node.children.len() {
+                // Visit the child by pushing it to the stack.
+                // Also increase `nth_child` so that when we come back to the `tag` node, we
+                // look at the next child.
+                let next_child = node.children[*nth_child];
+                *nth_child += 1;
+                stack.push((next_child, 0));
+                continue;
+            } else {
+                // First, get the current children of `node`.
+                mem::swap(&mut self.nodes.get_mut(*tag).unwrap().children, &mut children_of_node);
```

Comment:
This is probably more elegant as `let mut children_of_node = mem::take(&mut ...);`